src/sympc/protocol/aby3/aby3.py

```diff
@@ -92,10 +93,25 @@ def truncation_algorithm1(
         base = config.encoder_base
         precision = config.encoder_precision
         scale = base ** precision
+
         x1, x2, x3 = ptr_list
+        # TODO: UNSIGNED Truncation does not work for numpy, should be modified.
+        dtype = ptr_list[0].dtype
+        if isinstance(ptr_list[0], np.ndarray):
```

Comment:
Q: Shouldn't it work - if you have a uint64 numpy array type and you want to add another np.uint array to it, it should wrap around, right?
I think the problem comes when you add an integer, and at that point the numpy type gets converted to an `object` type.

Reply:
Yes, George, it should wrap around, I am pretty sure the truncation algorithm should work for unsigned types, it needs additional look into  the implementation semantics.I will try figuring out the problem.

Couldn't get you  , the `object` type meaning?

Reply:
Hmm..a first question  - from where does the unsigned comes? (the unsigned numpy)

Reply:
Crct , the error occurs for unsigned numpy types.

Reply:
No no, my question was more. From where do we reach a possible `numpy` array here?

Reply:
Oh ok , In Private Compare,we convert torch tensor to numpy. We invoke a multiplication operation ,it reaches here for truncation george.

Reply:
At a high level  ReLU -> Wrap -> PC in the order of invocation . ReLU works with torch tensors,At wrap we convert to Numpy.PC does some additional conversion in Numpy for random generation.I think in Numpy we cannot generate random tensor with custom shape like we do in torch(It has for floats, but not integers),we use torch random tensor genearation , and convert to numpy.

Reply:
Hmm...but why don't we convert back to tensor in PC?

Reply:
PC is a subroutine in wrap george, I left it like that , as we require numpy there.

Reply:
yep, but it does not ask my question. but if we convert there back to tensor there, we won't have all the logic where we check if we have a numpy array (in all the other places), right?

Reply:
ping :eyes:

Reply:
I think not george, for example in the wrap pre processing step `function` we use all the primitive operations of MPC (add,sub,mul),so we had to include support for numpy and a corresponding sanity check for the same.Was this the question, I am not sure , If I answered the question.

Reply:
By all other places  , you meant the (`ABY,addmodprime,submodprime,mulmodprime`) crct george?

Reply:
So my intention was to remove the numpy logic and convert to ShareTensor/Tensor where we have the numpy array - from my understanding we convert to numpy array only in PC, right (as per comment
"Oh ok , In Private Compare,we convert torch tensor to numpy. We invoke a multiplication operation in the subsequent operation in Private Compare ,it reaches here for truncation george." -- and in this case can't we convert back to tensor in PC such that we won't have all the numpy logic in all the places?)

Reply:
Sorry , My bad, As the comment was placed in ABY3, I was answering in that context.
I will give a comprehensive explanation.
We use `numpy` in wrap context mainly.
Why are we doing that?To suppose the claim, Falcon wrap is designed for `unsigned` numbers , What we really want to do is check for carry using unsigned numbers.

Reply:
http://teaching.idallen.com/dat2343/10f/notes/040_overflow.txt , this article helped me in understanding the subtleties of overflow and underflow

Reply:
Since we use in numpy in wrap and PC is a subroutine of Wrap, it is also made compatible for numpy,

Reply:
The main reason , why we modify all the operations , in many places, is that , when we use numpy in wrap, it undergoes lot of primitive operations of MPC , like the (add, sub, mul) , so we make it compatible with all ops.

Reply:
https://github.com/OpenMined/SyMPC/blob/5b699405716a0e1999b0b84efb8e4a84ca1b7c99/src/sympc/protocol/falcon/falcon.py#L647 I think this function of wrap , would help to understand

Reply:
I will have a look at the wrap today - I am not sure why we are not able to use the carry bit if we have signed numbers (under the hood - the bit representation should be the same for signed/unsigned numbers -- only that the unsigned numbers have 1 more bit for the representation which for the signed numbers that one is the sign bit)